src/details/ast.hpp

```diff
@@ -19,74 +20,84 @@ class FieldExpression;
 
+enum GEN_CAUSE { FOR_LOOP_DESUGAR, FOR_IN_LOOP_DESUGAR, FUNCTION_STATEMENT_DESUGAR };
 enum class LiteralType { TRUE, FALSE, NIL, NUMBER, STRING };
 class Literal {
     std::string literal_content;
     LiteralType literal_type;
-    ts::Range literal_range;
+    minilua::Range literal_range;
 
 public:
-    Literal(LiteralType, std::string, ts::Range);
+    Literal(LiteralType, std::string, minilua::Range);
     auto content() const -> std::string;
     auto type() const -> LiteralType;
     auto range() const -> minilua::Range;
 };
 /**
- * The Body class groups a variable amount of statements together
- * the last statement of a Body might be a return_statement
+ * class for program nodes
  */
-class Body {
-    std::vector<ts::Node> nodes;
+class Program {
+    ts::Node program;
 
 public:
-    explicit Body(std::vector<ts::Node>);
-    /**
-     * This method maps the statement Nodes to Statement classes
-     * @return the statements in this body excluding a potential return_statement
-     */
-    auto statements() -> std::vector<Statement>;
+    explicit Program(ts::Node);
     /**
-     * This checks for a return node
-     * the return_statement is always the last statement of the body
-     * @return a Return class if the body has a return statement
-     *          else an empty optional
+     * the children nodes of the program get put into a Body class by this method
+     * @return a Body containing the full program
      */
-    auto return_statement() -> std::optional<Return>;
+    auto body() const -> Body;
+    auto range() const -> minilua::Range;
+    auto debug_print() const -> std::string;
 };
 /**
- * class for program nodes. It only holds a body
+ * class for do_statement nodes
  */
-class Program {
-    ts::Node program;
+class DoStatement {
+    struct DoStruct {
+        std::shared_ptr<Body> body; // the pointer is needed because Body is only a forward
+                                    // declaration here and no complete type
+        minilua::Range range;
+        GEN_CAUSE gen_cause;
+        DoStruct(const Body&, minilua::Range, GEN_CAUSE);
+    };
+    std::variant<ts::Node, DoStruct> content;
 
 public:
-    explicit Program(ts::Node);
+    explicit DoStatement(ts::Node);
+    explicit DoStatement(const Body&, minilua::Range, GEN_CAUSE);
     /**
-     * the children nodes of the Program get put into a Body class by this method
-     * @return a Body containing the full program
+     *
+     * @return a body containing all statements of the do block
      */
     auto body() const -> Body;
     auto range() const -> minilua::Range;
-    auto raw() const -> ts::Node;
+    auto debug_print() const -> std::string;
 };
 /**
  * class for identifier_nodes
  */
 class Identifier {
-    ts::Node id;
+    struct IdStruct {
+        std::string identifier;
+        minilua::Range range;
+        GEN_CAUSE gen_cause;
+        IdStruct(std::string, minilua::Range, GEN_CAUSE);
```

Comment:
This is fine. But you don't necessarily need a constructor here. You could just initialize it with `IdStruct{ id, range, cause };`.